app/com/arpnetworking/metrics/portal/query/impl/KairosDbQueryExecutor.java

```diff
@@ -120,12 +122,27 @@ public Duration lookbackPeriod(final MetricsQuery query) {
             throw new RuntimeException("Could not parse query", e);
         }
         // The lookback period of the query is the largest of each metric within
-        return metricsQuery.getMetrics()
+        final Duration period = metricsQuery.getMetrics()
                 .stream()
                 .map(this::lookbackPeriod)
                 .flatMap(Streams::stream)
                 .max(Duration::compareTo)
                 .orElseThrow(() -> new IllegalArgumentException("Query did not specify any range aggregators"));
+
+        return new DefaultQueryWindow.Builder()
+                .setPeriod(period)
+                .setAlignment(getAlignment(metricsQuery))
+                .build();
+    }
+
+    private QueryAlignment getAlignment(final com.arpnetworking.kairos.client.models.MetricsQuery metricsQuery) {
+        final boolean anyEndAligned = metricsQuery.getMetrics()
+                .stream()
+                .flatMap(m -> m.getAggregators().stream())
+                .flatMap(agg -> Streams.stream(agg.getAlignEndTime()))
+                .anyMatch(endAligned -> endAligned);
```

Comment:
If it makes more sense, I can validate the assumption here that either all aggregators are end aligned or all aggregators are period aligned.

This isn't enforced by the data model, but the windowing logic relies on this. A mixed set of alignment would require the query window to start at the same time as period aligned and end at the same time as end-aligned.